### ISSUE TITLE
splatting in tuple construction

### ISSUE THREAD

**Issue by stevengj** [JuliaLang/julia#4869]
I was surprised to discover that this doesn't work:
```
(3,4,(5,6)...)
```
I was expecting `(3,4,5,6)`, but got `type: apply_type: in Vararg, expected Type{T<:Top}, got (Int64,Int64)`

Even more surprising, 
```
(3,4,()...)
```
works, but outputs `(3,4,()...)` instead of `(3,4)`, with `(3,4,()...)[3]` being a somewhat mysterious `()...` object of type `DataType`.

I understand that some special magic has to occur here in order to be able to define tuple types like `(Int, Bool, Float64...)`, but the behavior above seems odd.

**Comment by JeffBezanson** [comment]
`{...Int}` is a good idea but it looks a little weird to me. Instead I'll propose `{Int, ...}`. That also has the interesting property that you can write `{Xs..., ...}`, which splats but also makes the type vararg by repeating the last element of `Xs`.

**Comment by nalimilan** [comment]
I agree that @JeffBezanson's proposal looks more natural.

**Comment by StefanKarpinski** [comment]
So in the proposed syntax `{Int, ...}` is the type of `Int` tuples of zero or greater? Whereas `{Int...}` would attempt to splat `Int` when constructing a tuple type object?

**Comment by mbauman** [comment]
I like `{Int, ...}`, but it seems to strongly suggest that the first element would be required (i.e., one-or-more ints).

**Comment by StefanKarpinski** [comment]
Despite the nice analogy between `{A,B}` is to `T{A,B}` as `(a,b)` is to `f(a,b)`, I'm still not entirely convinced that we need such a convenient syntax for tuple types.

**Comment by JeffBezanson** [comment]
It could become more important in the future, e.g. with function types `{A,B}->C`. It also opens up other syntactic space for types. We could consider something like `T.{Array{T}}` for "UnionAll" types.

@mbauman good point that `{Int, ...}` looks like "one or more". I think I can live with that though.

**Comment by JeffBezanson** [comment]
Actually the syntax for UnionAll types could be `{T<:S}(Array{T})`, exactly how they're currently written for method signatures.

**Comment by StefanKarpinski** [comment]
Using parens in `{T<:S}(Array{T})` seems inconsistent with using `{...}` for tuple types. The prospect of making this work nicely for existential union types is appealing. That's something we *do* very much need a good syntax for.

**Comment by JeffBezanson** [comment]
Yes, I suppose it conflicts with the ability to "call" a tuple type. This currently works:

```
julia> (Int,Int)((1.0,2.0))
(1,2)
```

`{T<:S}{Array{T}}` is also available for existentials though.

Going back to function types briefly, if we really want I guess we could have `{A,B}->C` but use `Tuple{A,B}` for tuples otherwise.

**Comment by StefanKarpinski** [comment]
We could also use `(A,B)-->C` as planned in ancient history.

**Comment by JeffBezanson** [comment]
The downside is that it wouldn't fit with writing `T-->C` where `T` is a tuple type already. Actually that's also why we'd need `-->` and couldn't use `->` as I've been writing. But that part's easy.

**Comment by StefanKarpinski** [comment]
I don't get that bit – can you elaborate?

**Comment by JeffBezanson** [comment]
You want to be able to do

```
T = (A,B)
ftype = T-->C
```

so that there can be a single variable that represents all of the arguments. You want to be able to pull out any subexpression into a variable. Of course it's *possible* to have `(A,B)-->C` be equivalent to

```
T = Tuple{A,B}
T-->C
```

but I think pretty suboptimal.

**Comment by mauro3** [comment]
How does this fit with the discussion over in #6984 where `{T<:S}->Array{T}` was suggested for what Jeff wrote above `{T<:S}(Array{T})`?

**Comment by JeffBezanson** [comment]
fixed by #10380
